server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CounterResource.java

```diff
@@ -0,0 +1,247 @@
+            resourceRegistration.registerOperationHandler(CounterResource.COUNTER_DECREASE,
+                    CounterDecreaseCommand.INSTANCE);
+        }
+    }
+
+    private static SimpleOperationDefinitionBuilder buildOperation(String name) {
+        return new SimpleOperationDefinitionBuilder(name, new InfinispanResourceDescriptionResolver(ModelKeys.COUNTERS))
+                .setRuntimeOnly();
+    }
+
+    private static PathElement counterElement(OperationContext context, ModelNode operation)
+            throws OperationFailedException {
+        final PathAddress address = pathAddress(operation.require(OP_ADDR));
+        final PathElement counterElement = address.getElement(address.size() - 1);
+        return counterElement;
+    }
+
+    private static String counterName(OperationContext context, ModelNode operation) throws OperationFailedException {
+        PathElement counterElement = counterElement(context, operation);
+        return counterElement.getValue();
+    }
+
+    private static String counterType(OperationContext context, ModelNode operation) throws OperationFailedException {
+        PathElement counterElement = counterElement(context, operation);
+        return counterElement.getKey();
+    }
+
+    private static OperationFailedException counterManagerNotFound() {
+        return new OperationFailedException("CounterManager not found in server.");
+    }
+
+    public static class CounterRemoveHandler extends AbstractRemoveStepHandler {
+
+        private final CounterAddHandler handler;
+
+        public CounterRemoveHandler() {
+            handler = new CounterAddHandler();
+        }
+
+        @Override
+        protected void performRuntime(final OperationContext context, final ModelNode operation, final ModelNode model) throws OperationFailedException {
+            handler.removeRuntimeServices(context, operation, null, null);
+        }
+
+        @Override
+        protected void recoverServices(final OperationContext context, final ModelNode operation, final ModelNode model) throws OperationFailedException {
+            handler.performRuntime(context, operation, model);
+        }
+    }
+
+    private static class CounterResetCommand extends BaseCounterManagerCommand {
+        private static final CounterResetCommand INSTANCE = new CounterResetCommand();
+
+        @Override
+        protected ModelNode invoke(CounterManager counterManager, OperationContext context, ModelNode operation)
+                throws Exception {
+            final String counterName = counterName(context, operation);
+            final String counterType = counterType(context, operation);
+            if (counterManager.isDefined(counterName)) {
+                boolean isStrongCounter = ModelKeys.STRONG_COUNTER.equals(counterType);
+                if (isStrongCounter) {
+                    StrongCounter strongCounter = counterManager.getStrongCounter(counterName);
+                    strongCounter.reset();
```

Comment:
the `reset()` is asynchronous! you are returning while the operation is executed. I'm not sure if it is on purpose or not. If not, use the `sync()` method.
The same note for the increment/decrement

Reply:
up ^